language/tools/move-mv-llvm-compiler/tests/rbpf-tests/moption01.move

```diff
@@ -0,0 +1,440 @@
+    public fun fill<Element>(t: &mut Option<Element>, e: Element) {
+        let vec_ref = &mut t.vec;
+        if (vector::is_empty(vec_ref)) vector::push_back(vec_ref, e)
+        else abort EOPTION_IS_SET
+    }
+
+    /// Convert a `some` option to a `none` by removing and returning the value stored inside `t`
+    /// Aborts if `t` does not hold a value
+    public fun extract<Element>(t: &mut Option<Element>): Element {
+        assert!(is_some(t), EOPTION_NOT_SET);
+        vector::pop_back(&mut t.vec)
+    }
+
+    /// Return a mutable reference to the value inside `t`
+    /// Aborts if `t` does not hold a value
+    public fun borrow_mut<Element>(t: &mut Option<Element>): &mut Element {
+        assert!(is_some(t), EOPTION_NOT_SET);
+        vector::borrow_mut(&mut t.vec, 0)
+    }
+
+    /// Swap the old value inside `t` with `e` and return the old value
+    /// Aborts if `t` does not hold a value
+    public fun swap<Element>(t: &mut Option<Element>, e: Element): Element {
+        assert!(is_some(t), EOPTION_NOT_SET);
+        let vec_ref = &mut t.vec;
+        let old_value = vector::pop_back(vec_ref);
+        vector::push_back(vec_ref, e);
+        old_value
+    }
+
+    /// Swap the old value inside `t` with `e` and return the old value;
+    /// or if there is no old value, fill it with `e`.
+    /// Different from swap(), swap_or_fill() allows for `t` not holding a value.
+    public fun swap_or_fill<Element>(t: &mut Option<Element>, e: Element): Option<Element> {
+        let vec_ref = &mut t.vec;
+        let old_value = if (vector::is_empty(vec_ref)) none()
+            else some(vector::pop_back(vec_ref));
+        vector::push_back(vec_ref, e);
+        old_value
+    }
+
+    /// Destroys `t.` If `t` holds a value, return it. Returns `default` otherwise
+    public fun destroy_with_default<Element: drop>(t: Option<Element>, default: Element): Element {
+        let Option { vec } = t;
+        if (vector::is_empty(&mut vec)) default
+        else vector::pop_back(&mut vec)
+    }
+
+    /// Unpack `t` and return its contents
+    /// Aborts if `t` does not hold a value
+    public fun destroy_some<Element>(t: Option<Element>): Element {
+        assert!(is_some(&t), EOPTION_NOT_SET);
+        let Option { vec } = t;
+        let elem = vector::pop_back(&mut vec);
+        vector::destroy_empty(vec);
+        elem
+    }
+
+    /// Unpack `t`
+    /// Aborts if `t` holds a value
+    public fun destroy_none<Element>(t: Option<Element>) {
+        assert!(is_none(&t), EOPTION_IS_SET);
+        let Option { vec } = t;
+        vector::destroy_empty(vec)
+    }
+
+    /// Convert `t` into a vector of length 1 if it is `Some`,
+    /// and an empty vector otherwise
+    public fun to_vec<Element>(t: Option<Element>): vector<Element> {
+        let Option { vec } = t;
+        vec
+    }
+}
+
+//-----------------------------------------------------------------------------
+
+//module std::vector {
+module 0x10::vector {
```

Comment:
It seems unfortunate that vector has to be defined in every test file.

Reply:
> > As a separate and orthogonal (and not terribly important) task, one could implement another test directive such as // include-file: foo.inc and update the test files accordingly.
> 
> Good point. i'll add a task for this.

FWIW, I guess even easier is to just have the test driver use an existing macro processor (e.g., cpp or m4) and use a `#include`. Or make each test needing those includes a simple self-creating `.move` file:
```
#!/bin/bash
f="foo.move"
echo -n "Creeating $f ... "
start_line_of_move_test=$((`grep -an "^MOVE_TEST_CONTENT$" $0 | cut -d: -f1` + 1))
cat vector.move.inc > $f
cat option.move.inc >> $f
tail -n+${start_line_of_move_test} $0 >> $f
echo "done"
exit 0
MOVE_TEST_CONTENT
script {
    fun main() {
        use 0x10::vector;

        // Very simple tests for some non-native vector functions with
        // element type u64.
        // This exercises generic function support combined with vectors.

        // Test non-native `singleton` and `is_empty`.
        let v = vector::singleton<u64>(123);
        assert!(vector::length(&v) == 1, 0xf00);
        assert!(!vector::is_empty<u64>(&v), 0xf01);
        assert!(vector::pop_back<u64>(&mut v) == 123, 0xf02);
        assert!(vector::is_empty<u64>(&v), 0xf03);
    }
}
```